src/adapter/src/client.rs

```diff
@@ -1032,3 +1039,62 @@ impl RecordFirstRowStream {
         msg
     }
 }
+
+/// ~~SPOOKY ZONE~~
+///
+/// There is a small race that can occur with [`tokio`], a [`SessionClient`], and the `Coordinator`
+/// that requires this Drop guard.
+///
+/// 1. A [`SessionClient`] can execute a command, sending it's owned [`Session`] to the
+///    Coordinator.
+/// 2. While processing this request, the user closes their connection to Materialize.
+/// 3. The Coordinator (which is run concurrently on a different thread and in a different task)
+///    finishes processing our request and is able to successfully send a response over the
+///    provided [`oneshot::Sender`].
+/// 4. [`tokio`] realizes the connection is closed and Drops the task which is awaiting the result
```

Comment:
Why can the `tx` send the command (with the session inside) without an error? Is it because the receiver hasn't quite been dropped yet (Or is in the process of being dropped?)

Reply:
Yeah, exactly. The `Drop` impl of `SessionClient` runs before the `Drop` impl of `Receiver` runs and closes the channel. And in general the only way to close a oneshot channel without losing the message inside is to call `close()` on it manually.

Reply:
Yeah, til about `close()`+`try_recv()`, which is quite nice!